search_util.go

```diff
@@ -0,0 +1,592 @@
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// AddDateTimeRange adds a bucket to a field
+// containing date values.  Documents with a
+// date value falling into this range are tabulated
+// as part of this bucket/range.
+func (fr *FacetRequest) AddDateTimeRange(name string, start, end time.Time) {
+	if fr.DateTimeRanges == nil {
+		fr.DateTimeRanges = make([]*dateTimeRange, 0, 1)
+	}
+	fr.DateTimeRanges = append(fr.DateTimeRanges, &dateTimeRange{Name: name, Start: start, End: end})
+}
+
+// AddDateTimeRangeString adds a bucket to a field
+// containing date values. Uses defaultDateTimeParser to parse the date strings.
+func (fr *FacetRequest) AddDateTimeRangeString(name string, start, end *string) {
+	if fr.DateTimeRanges == nil {
+		fr.DateTimeRanges = make([]*dateTimeRange, 0, 1)
+	}
+	fr.DateTimeRanges = append(fr.DateTimeRanges,
+		&dateTimeRange{Name: name, startString: start, endString: end})
+}
+
+// AddDateTimeRangeString adds a bucket to a field
+// containing date values. Uses the specified parser to parse the date strings.
+// provided the parser is registered in the index mapping.
+func (fr *FacetRequest) AddDateTimeRangeStringWithParser(name string, start, end *string, parser string) {
+	if fr.DateTimeRanges == nil {
+		fr.DateTimeRanges = make([]*dateTimeRange, 0, 1)
+	}
+	fr.DateTimeRanges = append(fr.DateTimeRanges,
+		&dateTimeRange{Name: name, startString: start, endString: end, DateTimeParser: parser})
+}
+
+// AddNumericRange adds a bucket to a field
+// containing numeric values.  Documents with a
+// numeric value falling into this range are
+// tabulated as part of this bucket/range.
+func (fr *FacetRequest) AddNumericRange(name string, min, max *float64) {
+	if fr.NumericRanges == nil {
+		fr.NumericRanges = make([]*numericRange, 0, 1)
+	}
+	fr.NumericRanges = append(fr.NumericRanges, &numericRange{Name: name, Min: min, Max: max})
+}
+
+// FacetsRequest groups together all the
+// FacetRequest objects for a single query.
+type FacetsRequest map[string]*FacetRequest
+
+func (fr FacetsRequest) Validate() error {
+	for _, v := range fr {
+		if err := v.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// HighlightRequest describes how field matches
+// should be highlighted.
+type HighlightRequest struct {
+	Style  *string  `json:"style"`
+	Fields []string `json:"fields"`
+}
+
+// NewHighlight creates a default
+// HighlightRequest.
+func NewHighlight() *HighlightRequest {
+	return &HighlightRequest{}
+}
+
+// NewHighlightWithStyle creates a HighlightRequest
+// with an alternate style.
+func NewHighlightWithStyle(style string) *HighlightRequest {
+	return &HighlightRequest{
+		Style: &style,
+	}
+}
+
+func (h *HighlightRequest) AddField(field string) {
+	if h.Fields == nil {
+		h.Fields = make([]string, 0, 1)
+	}
+	h.Fields = append(h.Fields, field)
+}
+
+func (r *SearchRequest) Validate() error {
```

Comment:
@metonymic-smokey Let's add a Validate() method for the `knn` part of the search request and call it here when applicable.